Allow setting Kyverno enforcement

buraksekili opened the pull request:
**What this PR does / why we need it**:

This PR implements hierarchical Kyverno enforcement at Datacenter, Seed, and Global (KubermaticConfiguration) levels. This allows KKP administrators to mandate Kyverno deployment on user clusters at different scopes.

**Which issue(s) this PR fixes**:
<!--optional, in `fixes #<issue number>` format, will close the issue(s) when PR gets merged-->
Fixes #14850

**What type of PR is this?**
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
/kind chore
-->
/kind feature


**Special notes for your reviewer**:

- When `enforced: true` at any level => Kyverno is mandatory for clusters under that scope
- When `enforced: false` at any level =? Kyverno is explicitly not enforced at that level
- When `enforced: nil `(not set) => No opinion, check next level in hierarchy
Higher levels override lower levels (DC > Seed > Global)

Test cases can be found here: pkg/util/kyverno/kyverno_test.go, as a reference for regression.


**Does this PR introduce a user-facing change? Then add your Release Note here**:
<!--
Write your release note. Release notes are being used to generate the changelog:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Add support for Kyverno enforcement at Datacenter, Seed, and Global levels. Administrators can now enforce Kyverno policy engine deployment on user clusters with configurable precedence (Datacenter > Seed > Global). When enforced, Kyverno is automatically enabled for clusters.
```

**Documentation**:
<!--
Please do one of the following options:
- Add a link to the existing documentation
- Add a link to the kubermatic/docs pull request
- If no documentation change is applicable then add:
  - TBD (documentation will be added later)
  - NONE (no documentation needed for this PR)
-->
```documentation
TBD
```

buraksekili committed:
feat: Allow setting Kyverno enforcement

Implement hierarchical enforcement for Kyverno at Datacenter, Seed, and KubermaticConfiguration (Global) levels to address issue #14850.
When enforced at any level, Kyverno is automatically enabled for user clusters and cannot be disabled at user cluster level.

- Add KyvernoConfigurations to Datacenter, Seed, and KubermaticConfiguration specs
- Implement GetKyvernoEnforcement with precedence: DC > Seed > Global
- Add validation to prevent users from disabling enforced Kyverno
- Add mutation to auto-enable Kyverno on cluster creation when enforced
- Update Kyverno controller to watch Seed/Config changes and enforce settings
- Add unit tests for enforcement logic
- Update CRDs with new Kyverno configuration fields

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

buraksekili committed:
fix linter

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

adoi commented on sdk/apis/kubermatic/v1/datacenter.go:
```diff
@@ -323,6 +323,23 @@ type SeedSpec struct {
+	// must have Kyverno enabled and cannot disable it.
+	// If it is set to true, the Kyverno becomes enforced at this level, and will be deployed to user clusters under the scope.
+	// If it is set to false, Kyverno is not enforced at this level.
+	// If nil, Kyverno no enforcement preference is declared.
```
small typo, should be`If nil, no Kyverno enforcement preference is declared`.

buraksekili replied:
addressed in dea93419fa8ad84965830abde42e4523bad56d22

adoi commented on sdk/apis/kubermatic/v1/datacenter.go:
```diff
@@ -323,6 +323,23 @@ type SeedSpec struct {
+	// If it is set to true, the Kyverno becomes enforced at this level, and will be deployed to user clusters under the scope.
+	// If it is set to false, Kyverno is not enforced at this level.
+	// If nil, Kyverno no enforcement preference is declared.
+	// For example, if Kyverno is enforced at the Seed level, setting this to false at the Datacenter level make the Datacenter not enforce Kyverno,
```
`... setting this to false at the Datacenter level makes the Datacenter not enforce Kyverno,`

adoi commented on pkg/mutation/cluster/mutation.go:
```diff
@@ -47,6 +48,20 @@ func MutateCreate(newCluster *kubermaticv1.Cluster, config *kubermaticv1.Kuberma
 		newCluster.Spec.DisableCSIDriver = true
 	}
+
+	enforcementInfo := util.GetKyvernoEnforcement(
```
I would add the enforcement logic in `MutateUpdate` too (please check my comment for putting it in DefaultClusterSpec too), so that we don't rely on controller timing and avoid potential race conditions. 

If you would prefer to keep it like this (without MutateUpdate logic) and trust the controller, I would still relax the logic in `validateKyvernoEnforcement` to only block explicit disable attempts, so that users would be able to update clusters during a race window.

The `validateKyvernoEnforcement` at the moment doesn't check *what* the user is trying to update (in our case we should only care about old vs new cluster, and if Kyverno was enabled and it's trying to be disabled we reject it).

adoi commented on sdk/apis/kubermatic/v1/datacenter.go:
```diff
@@ -323,6 +323,23 @@ type SeedSpec struct {
+	// Kyverno configures the Kyverno policy engine settings at the seed level.
+	// These settings apply to all user clusters in this seed.
+	// +optional
+	Kyverno *KyvernoConfigurations `json:"kyverno,omitempty"`
```
Is there any particular reason the logic wasn't added to DefaultClusterSpec? 

I would consider moving the Kyverno enforcement logic from MutateCreate to DefaultClusterSpec. Since DefaultClusterSpec runs on both create and update, this would automatically enforce Kyverno on existing clusters when they're next updated and eliminate the race condition between enabling enforcement and the controller patching clusters. This also matches the existing patterns i.e for KubeLB enforcement.

buraksekili replied:
Makes sense. updated in dea93419fa8ad84965830abde42e4523bad56d22

buraksekili committed:
Relocated Kyverno enforcement resolution from pkg/controller/util to a new pkg/util/kyverno package for better organization.

Also moved the enforcement defaulting from the mutation webhook to pkg/defaulting/cluster.go for consistency.
The validation webhook now allows updates when Kyverno was already enabled before enforcement was configured, preventing issues with existing clusters.

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

buraksekili committed:
add docs

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

buraksekili committed:
fix generated docs

Signed-off-by: Burak Sekili <32663655+buraksekili@users.noreply.github.com>

buraksekili committed:
Merge branch 'main' into feat/kyverno-enforcement

adoi reviewed (approved):
Great job! 

/approve
/lgtm
